packages/provider-elasticsearch/src/example-types/results/highlighting/response.js

```diff
@@ -0,0 +1,189 @@
+import _ from 'lodash/fp.js'
+import F from 'futil'
+import { isArraysEqual, groupByIndexed } from '../../../utils/futil.js'
+import {
+  stripTags,
+  mergeHighlights,
+  isBlobField,
+  isArrayField,
+  isArrayOfScalarsField,
+  findByPrefix,
+  getArrayOfObjectsPathsMap,
+} from './util.js'
+
+let lastWordRegex = /\.(\w+)$/
+
+/*
+ * Group highlight results by their multifield. For example `city` and
+ * `city.subfield` will be grouped under `city`.
+ */
+let groupByMultiField = _.curry((schema, highlight) =>
+  groupByIndexed((v, path) => {
+    let [multi, sub] = path.split(lastWordRegex)
+    return schema.fields[multi]?.elasticsearch?.mapping?.fields?.[sub]
+      ? multi
+      : path
+  }, highlight)
+)
+
+/**
+ * Mutate hit `highlight`:
+ * 1. Fragments for text fields get merged.
+ * 2. Fragments for large (blob) text fields get concatenated.
+ * 3. Fragments for arrays get ordered based on the source array
+ */
+export let transformResponseHighlight = (
+  schema,
+  hit,
+  tags,
+  nestedArrayIncludes = {}
+) => {
+  let arrayOfObjectsPaths = _.keys(getArrayOfObjectsPathsMap(schema))
+
+  let getIndexedArrayObject = (arr, fragments, arrayPath, itemPath) => {
+    let fragmentsMap = _.groupBy(stripTags(tags), fragments)
+    return F.reduceIndexed(
+      (acc, item, index) => {
+        let fragments = fragmentsMap[itemPath ? _.get(itemPath, item) : item]
+        if (fragments) {
+          F.setOn(
+            F.dotJoin([`${index}`, itemPath]),
+            mergeHighlights(tags, ...fragments),
+            acc
+          )
+          for (let itemPath of nestedArrayIncludes[arrayPath] ?? []) {
+            F.updateOn(
+              F.dotJoin([`${index}`, itemPath]),
+              (highlight) => highlight ?? _.get(itemPath, item),
+              acc
+            )
+          }
+        }
+        return acc
+      },
+      arr,
+      _.get(arrayPath, hit._source)
+    )
+  }
+
+  let getArrayPath = (path) =>
+    isArrayOfScalarsField(schema.fields[path])
+      ? path
+      : findByPrefix(path, arrayOfObjectsPaths)
+
+  let highlight = _.flow(
+    groupByMultiField(schema),
+    _.mapValues(_.flatten),
+    F.reduceIndexed((acc, fragments, path) => {
+      let arrayPath = getArrayPath(path)
+      if (arrayPath) {
+        acc[arrayPath] = getIndexedArrayObject(
+          acc[arrayPath] ?? {},
+          fragments,
+          arrayPath,
+          path.slice(arrayPath.length + 1)
+        )
+      } else if (isBlobField(schema.fields[path])) {
+        acc[path] = fragments
+      } else {
+        acc[path] = mergeHighlights(tags, ...fragments)
+      }
+      return acc
+    }, {})
+  )(hit.highlight)
+
+  if (!_.isEmpty(highlight)) hit.highlight = highlight
+}
+
+/**
+ * Remove each path in `paths` from `hit._source`.
+ */
+export let removePathsFromSource = (schema, hit, paths) => {
```

Comment:
Based on our discussion, you can likely substantially simplify this by thinking of it in terms of an `omit`. To handle the case that made this complicated (omitting all paths from an array of objects), you can determine that up front before iterating to omit.

The idea here is that you're reframing the problem up front to make the subsequent code much simpler.

Reply:
Reworked the code and now it's simpler